### PR TITLE
Streamwrapper: Cache file_exists requests

### PR DESCRIPTION
**Pull request by hanifn**
## Description

Start caching `file_exists()`/`stat()` requests to reduce number requests to the Fileservice.

Some actions like image editing can result in multiple `file_exists()` calls which may lead to timeouts on high traffic sites.

## Checklist

Please make sure the items below have been covered before requesting a review:

- [x] This change works and has been tested locally (or has an appropriate fallback).
- [x] This change works and has been tested on a Go sandbox.
- [x] This change has relevant unit tests (if applicable).

## Steps to Test

1. Set `VIP_FILESYSTEM_STREAM_WRAPPER_DEBUG` constant as `true` on a test site.
1. Do some media action i.e.: upload file, edit image etc.
1. Check number of `file_exists` requests ( Has `X-action: file_exists` header )
1. Checkout PR and repeat media actions
1. Verify number of `file_exists` requests is reduced.

### CODE REVIEW

File: files/class-api-cache.php [hunk 1]
```diff
@@ -40,14 +45,17 @@ public static function get_instance() {
...
 			return;
 		}
 
 		foreach( $this->files as $name => $path ) {
 			unlink( $path );
 			unset( $this->files[ $name ] );
 		}
+
+    // empty file stats cache
+    $this->file_stats = [];
```
Comment (nickdaugherty) [inline files/class-api-cache.php]: Looks like the whitespace on these lines got a little funky.

File: files/class-api-client.php [hunk 2]
```diff
@@ -219,6 +226,9 @@ public function is_file( $file_path, &$info = null ) {
 			$response_body = wp_remote_retrieve_body( $response );
 			$info = json_decode( $response_body, true );
 
+			// cache file info
+			$this->cache->cache_file_stats( $file_path, $info );
```
Comment (nickdaugherty) [inline files/class-api-client.php]: Would we also want to cache failures (404s), or is that overkill?

File: files/class-api-cache.php (hunk 1 above)
Comment (nickdaugherty) [reply files/class-api-cache.php]: Actually, I think most of the new lines of code have funky whitespace.

File: files/class-api-cache.php [hunk 3]
```diff
@@ -95,5 +120,10 @@ public function remove_file( $filepath ) {
 			unlink( $this->files[ $file_name ] );
 			unset( $this->files[ $file_name ] );
 		}
+
+		// Remove cached stats too if any
+		if ( isset( $this->file_stats[ $file_name ] ) ) {
```
Comment (nickdaugherty) [inline files/class-api-cache.php]: Technically we can omit the `isset()` check here for simplicity b/c PHP won't raise a warning if you `unset()` an undefined variable.

File: files/class-api-cache.php [hunk 4]
```diff
@@ -95,5 +120,10 @@ public function remove_file( $filepath ) {
...
+			unset( $this->file_stats[ $file_name ] );
```
Comment (nickdaugherty) [inline files/class-api-cache.php]: Should this be encapsulated into a reusable function, like we have for `cache_file_stats` and `get_file_stats`?

File: files/class-api-client.php (hunk 2 above)
Comment (hanifn) [reply files/class-api-client.php]: I think it's overkill. I can't think of a situation where that would be necessary.

File: files/class-api-cache.php (hunk 4 above)
Comment (hanifn) [reply files/class-api-cache.php]: I'm not sure if that's necessary. It will basically be a one line function wouldn't it?

**Comment by hanifn** [comment]
> It would also be good to add a test case to verify that calling API_Cache::remove_file() also clears the stat cache.

Added the test in 37373212202674e2c021e8994ea2034c67d1cac7

File: files/class-api-cache.php (hunk 1 above)
Comment (hanifn) [reply files/class-api-cache.php]: Ah looks like vim isn't remembering my `noexpandtab` setting for some reason 😐

Fixed in 6ec94e12a2fc368541ee9e57e0d236630e4033a7

File: files/class-api-cache.php (hunk 3 above)
Comment (hanifn) [reply files/class-api-cache.php]: Good point. Changed in 990f7cd84cc27773ecc94026cdb9773cf77821e8

**Comment by chrean** [comment]
> This is a good start - are there other scenarios where we need to clear the cache though?

> Specifically, I'm thinking if a file is modified during the request - I don't see where that causes the stats cache to be cleared (maybe missing it).

Looks good @hanifn, my only concern is the same as @nickdaugherty raised just above.
I think we should invalidate/clear the cache whenever a file is modified otherwise the system may encounter inconsistencies and this will bite us back in the future.

**Comment by hanifn** [comment]
>  I think we should invalidate/clear the cache whenever a file is modified otherwise the system may encounter inconsistencies and this will bite us back in the future.

Good point. Changed in e4d80d0bd6903b04e1342cd33d4f074b19ed84af

**Comment by hanifn** [comment]
r731-stacks

**Pull request merged**